### PR TITLE
[CU-feu7x7] saturn labels

### PR DESCRIPTION
**Pull request by bhperry**
- [x] passes `make lint`
- [x] adds tests to `tests/` (if appropriate)

## What does this PR change?

Automatically add saturn-specific labels to the flow environment. 

## How does this PR improve `prefect-saturn`?

Along with the related change in saturn, this allows flows to be assigned to the correct cluster agent even if there are agents in multiple clusters that are all using the same prefect-cloud tenant.

### CODE REVIEW

**Comment by hhuuggoo** [comment]
Task linked: [CU-feu7x7 [saturn][prefect-saturn] Add BASE_URL label on agents and flows](https://app.clickup.com/t/feu7x7)

File: prefect_saturn/core.py [hunk 1]
```diff
@@ -303,6 +301,7 @@ def _get_environment(
                 adapt_kwargs=adapt_kwargs,
             ),
             job_spec_file=local_tmp_file,
+            labels=[str(urlparse(self._base_url).hostname), "saturn-cloud"],
```
Comment (jameslamb) [inline prefect_saturn/core.py]: just want to note here that if this Prefect proposal to remove `KubernetesJobEnvironment` ever happens, we'll have to find a different place to put labels: https://github.com/PrefectHQ/prefect/issues/2928

but we'll have lots of other problems if that happens too so doesn't affect this PR :grinning:

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this! I left on the accompanying Atlas PR with some suggestions.

Have you checked what happens if someone passes labels into `flow.register()`? Do they override the set of labels from storage + environment, or do the labels you pass in get added to those from the storage + environment?

I'm trying to think through how we should handle existing customer code that is passing `labels` to `flow.register()`. I know we'll have to tell them to re-register their flows anyway, just want to be prepared for when someone inevitably misses also updating their `flow.register()` call.


One other request...can you please do a minor version bump in `VERSION.txt`? This is definitely a breaking change.

File: prefect_saturn/core.py (hunk 1 above)
Comment (bhperry) [reply prefect_saturn/core.py]: Oh fun. I kinda liked that we could have default labels on the environment, and then allow users to set their own on the flow (if we get around to user-defined labels on agents)

**Comment by bhperry** [comment]
> Have you checked what happens if someone passes labels into `flow.register()`? Do they override the set of labels from storage + environment, or do the labels you pass in get added to those from the storage + environment?
>
> I'm trying to think through how we should handle existing customer code that is passing `labels` to `flow.register()`. I know we'll have to tell them to re-register their flows anyway, just want to be prepared for when someone inevitably misses also updating their `flow.register()` call.

Yes, they get combined together. Which is very convenient for us if we want to allow users to set their own labels on agents and then specify them on a flow. So it won't be an issue if customers have `labels=["saturn-cloud"]` on their flow.register after the update.

> One other request...can you please do a minor version bump in `VERSION.txt`? This is definitely a breaking change.

I actually don't think this is a breaking change. Existing flows will continue to run as before (since the labels on agent are not required), and any flows created/updated won't need any code changes to continue working. They will just have better labels automatically applied.

**Comment by jameslamb** [comment]
I think it's a breaking change because if anyone on an older version of Saturn ~(with agents that only have labels `["saturn-cloud", "webhook-storage"]`)~ runs `pip install prefect-saturn` after we release this, then tries to register a new flow, ~their flow will no longer match their agents and won't run~ that code will break because labels won't be in the response from Atlas.

The only way for this to not be a breaking change is if we don't release it to PyPi until every customer has been upgraded and had the labels on their agents updated.

**Comment by bhperry** [comment]
> I think it's a breaking change because if anyone on an older version of Saturn ~(with agents that only have labels `["saturn-cloud", "webhook-storage"]`)~ runs `pip install prefect-saturn` after we release this, then tries to register a new flow, ~their flow will no longer match their agents and won't run~ that code will break because labels won't be in the response from Atlas.
> 
> The only way for this to not be a breaking change is if we don't release it to PyPi until every customer has been upgraded and had the labels on their agents updated.

That is true. Minor version bump it is.

File: prefect_saturn/core.py [hunk 2]
```diff
@@ -149,6 +147,7 @@ def _set_flow_metadata(self, flow: Flow) -> None:
         self._saturn_flow_id = str(response_json["id"])
         self._saturn_flow_version_id = response_json["flow_version_id"]
         self._saturn_image = response_json["image"]
+        self._saturn_flow_labels = response_json.get("labels", ["saturn-cloud"])
```
Comment (bhperry) [inline prefect_saturn/core.py:150]: Set this with a default to allow us to release prefect-saturn without worrying about customers that haven't yet been updated.

Comment (jameslamb) [reply prefect_saturn/core.py:150]: oooooo good idea! Seems so obvious now that I see it haha

Comment (jameslamb) [reply prefect_saturn/core.py:150]: now it can just be a patch version (`0.4.1`)

File: Makefile [hunk 3]
```diff
@@ -26,7 +26,7 @@ lint:
 
 .PHONY: unit-tests
 unit-tests:
-	pip uninstall -y prefect-saturn
+	pip uninstall -y prefect-saturn || true
```
Comment (jameslamb) [inline Makefile:29]: why was this necessary?

Comment (bhperry) [reply Makefile:29]: breaks when prefect-saturn isn't already installed

Comment (jameslamb) [reply Makefile:29]: that's not the behavior I've seen

image

Comment (jameslamb) [reply Makefile:29]: I mean, I'm fine with this change. But I'm just trying to understand why the behavior is different in your environment than it is in mind, in case I run into this in the future.

This is my `conda info`

```text
     active environment : None
            shell level : 0
       user config file : /home/jlamb/.condarc
 populated config files : 
          conda version : 4.8.4
    conda-build version : not installed
         python version : 3.7.6.final.0
       virtual packages : __glibc=2.27
       base environment : /home/jlamb/miniconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/jlamb/miniconda3/pkgs
                          /home/jlamb/.conda/pkgs
       envs directories : /home/jlamb/miniconda3/envs
                          /home/jlamb/.conda/envs
               platform : linux-64
             user-agent : conda/4.8.4 requests/2.22.0 CPython/3.7.6 Linux/5.4.0-52-generic ubuntu/18.04.4 glibc/2.27
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

And I have `pip 20.2.2`

Comment (bhperry) [reply Makefile:29]: mine is `9.0.1`. I also have python 2.7 as my system default, so `make unittest` actually doesn't work for me anyways. Happy to change it back if you prefer.

Comment (jameslamb) [reply Makefile:29]: haha you should definitely not be using `python` 2.7 to do any local development, since all of our code is 3.7+

You don't have to revert this change, it's harmless. But I strongly recommend putting a Python from conda onto your PATH and not using Python 2.7 for anything

Comment (bhperry) [reply Makefile:29]: Oh gods no, I don't _use_ it. It's just there :laughing: 

Using the default conda env is a good idea though

File: VERSION [hunk 4]
```diff
@@ -1 +1 @@
-0.4.0
+0.4.1
```
Comment (jameslamb) [inline VERSION]: can you the trailing newline that's currently in this file?

**Review by jameslamb — CHANGES_REQUESTED** [review]
looks awesome, thank you. I just left one very small request.

File: VERSION (hunk 4 above)
Comment (bhperry) [reply VERSION]: yep, done

**Review by jameslamb — APPROVED** [review]
Looks good! Since this was done in a non-breaking way, I think it can be released immediately.

You can follow the releasing steps here: https://github.com/saturncloud/prefect-saturn/blob/main/CONTRIBUTING.md#releasing. Ignore the "make a PR" part since this is that PR.

Please preserve the structure I have from other releases, with the broken-out sections

image

**Comment by bhperry** [comment]
Cool. I'd like to test this against the latest release first, but otherwise I agree it should be fine to push out.

**Comment by bhperry** [comment]
@jameslamb I'm not authorized to merge to main. Can you grant me permission and/or merge this for me?

**Comment by jameslamb** [comment]
Oh I realize I never hit "send" on the comment....yep I added you as a maintainer on the repo, sorry about that
